File: kotlin/kindergarten-garden/src/main/kotlin/KindergartenGarden.kt
class KindergartenGarden(val diagram: String) {

    init {
        require(diagram.all { it in "VRCG\n" })
    }

    private val rows = diagram.split('\n')
    init {
        require(rows.size == 2)
        require(rows[0].length == rows[1].length)
    }

    val students = listOf(
            "Alice", "Bob", "Charlie", "David", "Eve", "Fred",
            "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"
    )

    val plants = mapOf(
            'V' to "violets",
            'R' to "radishes",
            'C' to "clover",
            'G' to "grass"
    )

    val plots = rows[0].chunked(2)
                        .zip(rows[1].chunked(2))
                        .map { (r0, r1) ->
                            (r0 + r1).map { plants.getValue(it) }
                        }

    fun getPlantsOfStudent(student: String): List<String> {
        require(student in students)
        return plots[students.indexOf(student)]
    }
}

File: dart/kindergarten-garden/lib/kindergarten_garden.dart
enum Plant { radishes, clover, violets, grass }

enum Student {
  Alice, Bob, Charlie, David, Eve, Fred,
  Ginny, Harriet, Ileana, Joseph, Kincaid, Larry,
}

class KindergartenGarden {
  final String diagram;
  Map<Student, List<Plant>> plots = {};
  KindergartenGarden(this.diagram);

  List<Plant> plants(Student student) {
    if (plots.isEmpty) _parse();
    return plots[student]!;
  }

  void _parse() {
    var plantLookup = {
      'R': Plant.radishes,
      'C': Plant.clover,
      'V': Plant.violets,
      'G': Plant.grass
    };
    var lines = diagram.split('\n');

    var plants = List.generate((lines.first.length / 2).toInt(), (i) {
      var plot = lines.map((line) => line.substring(2 * i, 2 * i + 2)).join('');
      return plot.split('').map((p) {
        if (plantLookup[p] == null) throw ArgumentError('not a plant $p');
        return plantLookup[p]!;
      }).toList();
    });

    var students = Student.values;
    for (var i = 0; i < plants.length; i++) {
      plots[students[i]] = plants[i];
    }
  }
}
